File: 0085-maximal-rectangle/0085-maximal-rectangle.java
class Solution {
    private int f(int n, int[] nums) {
        int maxi = 0;
        Stack<Integer> st = new Stack<>();
        for (int i = 0; i <= n; i++) {
            while (!st.isEmpty() && (i == n || nums[st.peek()] >= nums[i])) {
                int height = nums[st.pop()];
                int width;
                if (st.isEmpty()) width = i;
                else width = i - st.peek() - 1;
                maxi = Math.max(maxi, height * width);
            }
            st.push(i);
        }
        return maxi;
    }
    
    public int maximalRectangle(char[][] matrix) {
        int n = matrix.length, m = matrix[0].length;
        int nums[] = new int[m], maxi = 0;
        for (int i = 0; i < n; i++) {
            for (int j = 0; j < m; j++) {
                if (matrix[i][j] == '1') nums[j] += 1;
                else nums[j] = 0;
            }
            maxi = Math.max(maxi, f(m, nums));
        }
        return maxi;
    }
}

File: 0085-maximal-rectangle/0085-maximal-rectangle.cpp
class Solution {
public:
    int largest_rectangle_in_histogram(vector<int>& nums) {
        int res = 0;
        stack<int> st;
        
        // for (auto& it : nums) cout << it << " ";
        // cout << "\n";
        
        for (int i = 0, n = nums.size(); i <= n; i++) {
            while (!st.empty() && (i == n || nums[st.top()] >= nums[i])) {
                int h = nums[st.top()]; st.pop();
                int w = st.empty() ? i : i - st.top() - 1;
                res = max(res, h * w);
            }
            st.push(i);
        }
        
        return res;
    }
    
    int maximalRectangle(vector<vector<char>>& matrix) {
        vector<int> nums(matrix[0].size());
        int res = 0;
        
        for (int i = 0, m = matrix.size(), n = matrix[0].size(); i < m; i++) {
            for (int j = 0; j < n; j++) {
                nums[j] = matrix[i][j] == '0' ? 0 : nums[j] + 1;
            }
            res = max(res, largest_rectangle_in_histogram(nums));
        }
        
        return res;
    }
};